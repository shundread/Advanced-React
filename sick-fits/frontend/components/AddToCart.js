import gql from "graphql-tag";
import { useMutation } from "@apollo/client";
import { CURRENT_USER_QUERY } from "../lib/useUser";

const ADD_TO_CART_MUTATION = gql`
  mutation ADD_TO_CART_MUTATION($id: ID!) {
    addToCart(productId: $id) {
      id
    }
  }
`;

export function AddToCart({ id }) {
  const [addToCart, { data, error, loading }] = useMutation(
    ADD_TO_CART_MUTATION,
    {
      variables: { id },
      refetchQueries: [{ query: CURRENT_USER_QUERY }],
    }
  );

  console.log(data, error, loading);
  return (
    <button type="button" onClick={addToCart} disabled={loading}>
      Add to cart üõíÔ∏è
    </button>
  );
}
