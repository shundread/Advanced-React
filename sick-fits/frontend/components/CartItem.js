import styled from "styled-components";
import { formatMoney } from "../lib/formatMoney";

const CartItemStyles = styled.li`
  border-bottom: 1px solid var(--lightGrey);
  display: grid;
  grid-template-columns: auto 1fr auto;
  padding: 1rem 0;
  img {
    margin-right: 1rem;
  }
  h3,
  p {
    margin: 0;
  }
`;

export function CartItem({ cartItem }) {
  const { product } = cartItem;
  if (!product) {
    console.error("Unable to find product from cartItem", cartItem);
    return null;
  }
  return (
    <CartItemStyles>
      <img
        width={100}
        src={product.photo?.src}
        alt={product.photo?.alt || product.name}
      />
      <div>
        <h3>{product.name}</h3>
        <p>
          {formatMoney(product.price * cartItem.quantity)}
          {" - "}
          <em>
            {cartItem.quantity} &times; {formatMoney(product.price)} each
          </em>
        </p>
      </div>
    </CartItemStyles>
  );
}
