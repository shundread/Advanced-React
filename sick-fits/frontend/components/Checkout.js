import styled from "styled-components";
import nProgress from "nprogress";
import gql from "graphql-tag";
import { useMutation } from "@apollo/client";
import { loadStripe } from "@stripe/stripe-js";
import {
  CardElement,
  Elements,
  useElements,
  useStripe,
} from "@stripe/react-stripe-js";
import { useState } from "react";

import { SickButton } from "./styles/SickButton";
import { ErrorMessage } from "./ErrorMessage";

const CREATE_ORDER_MUTATION = gql`
  mutation CREATE_ORDER_MUTATION($token: String!) {
    checkout(token: $token) {
      id
      charge
      total
      items {
        id
        name
      }
    }
  }
`;

const CheckoutFormStyles = styled.form`
  box-shadow = 0 1px 2px 2px rgba(0, 0, 0, 0.04);
  border: 1px solid rgba(0, 0, 0, 0.06);
  border-radius: 5px;
  padding: 1rem;
  display: grid;
  grid-gap: 1rem;
`;

const stripeLib = loadStripe(process.env.NEXT_PUBLIC_STRIPE_KEY);

export function Checkout() {
  return (
    <Elements stripe={stripeLib}>
      <CheckoutForm />
    </Elements>
  );
}

function CheckoutForm() {
  const [error, setError] = useState();
  const [loading, setLoading] = useState(false);
  const stripe = useStripe();
  const elements = useElements();
  const [checkout, { error: graphQLError }] = useMutation(
    CREATE_ORDER_MUTATION
  );

  // This wasn't working as a callback somehow, but maybe React was just
  // throwing a confusing error?
  async function handleSubmit(e) {
    // 1. Stop the form from submitting and turn the loader on
    e.preventDefault();
    console.log("We gotta deal with the payment now!");

    // 2. Start the page transition
    setLoading(true);
    nProgress.start();

    // 3. Create the payment method via stripe (Token caomes back here if successful)
    const { error, paymentMethod } = await stripe.createPaymentMethod({
      type: "card",
      card: elements.getElement(CardElement),
    });
    console.log("Payment method", paymentMethod);

    // 4. Handle any errors from stripe
    if (error) {
      console.error("Error is", error);
      nProgress.done();
      setError(error);
      setLoading(false);
      return;
    }

    // 5. Send the token from step 3 to our keystone server, via a custom mutation!
    const orderResult = await checkout({
      variables: {
        token: paymentMethod.id,
      },
    });

    console.log("Finished with the order");
    console.log(orderResult);

    // 6. Change the page to view the order
    // 7. Close the cart
    // 8. Turn the loader off
    nProgress.done();
    setLoading(false);
  }

  return (
    <CheckoutFormStyles onSubmit={handleSubmit}>
      <ErrorMessage error={error} />
      <ErrorMessage error={graphQLError} />
      <CardElement />
      <SickButton>Checkout Now</SickButton>
    </CheckoutFormStyles>
  );
}
