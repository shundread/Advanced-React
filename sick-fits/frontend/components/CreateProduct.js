import { useMutation } from "@apollo/client";
import gql from "graphql-tag";
import { useForm } from "../lib/useForm";
import { Form } from "./styles/Form";
import Router from "next/router";

import { ErrorMessage } from "./ErrorMessage";
import { ALL_PRODUCTS_QUERY } from "./Products";
import { PAGINATION_QUERY } from "./Pagination";

export const CREATE_PRODUCT_MUTATION = gql`
  mutation CREATE_PRODUCT_MUTATION(
    $name: String!
    $description: String!
    $price: Int!
    $image: String
  ) {
    createProduct(
      data: {
        name: $name
        description: $description
        price: $price
        status: "AVAILABLE"
        photo: { create: { src: $image, altText: $name } }
      }
    ) {
      id
      name
      price
      description
    }
  }
`;

const DefaultProduct = {
  image: "",
  name: "",
  price: 0,
  description: "",
};

export function CreateProduct() {
  const { inputs, clearForm, handleChange, resetForm } =
    useForm(DefaultProduct);

  const [createProduct, { loading, error }] = useMutation(
    CREATE_PRODUCT_MUTATION,
    {
      variables: inputs,
      refetchQueries: [
        { query: ALL_PRODUCTS_QUERY },
        { query: PAGINATION_QUERY },
      ],
    }
  );

  async function handleSubmit(event) {
    event.preventDefault();

    // Submit the input fields to the backend
    const { data, error } = await createProduct();
    if (error) return;
    clearForm();

    // Go to the product page
    Router.push({
      pathname: `/product/[id]`,
      query: { id: data.createProduct.id },
    });
  }

  return (
    <Form onSubmit={handleSubmit}>
      <ErrorMessage error={error} />
      <fieldset disabled={loading} aria-busy={loading}>
        <label htmlFor="image">
          Image
          <input
            required
            id="image"
            name="image"
            type="file"
            onChange={handleChange}
          />
        </label>
        <label htmlFor="name">
          Name
          <input
            required
            id="name"
            name="name"
            placeholder="Name"
            type="text"
            value={inputs.name}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="price">
          Price
          <input
            required
            id="price"
            name="price"
            placeholder="Price"
            type="number"
            value={inputs.price}
            onChange={handleChange}
          />
        </label>
        <label htmlFor="description">
          Description
          <textarea
            required
            id="description"
            name="description"
            placeholder="Description"
            value={inputs.description}
            onChange={handleChange}
          />
        </label>
      </fieldset>
      <button type="submit">+ Add Product</button>
    </Form>
  );
}
