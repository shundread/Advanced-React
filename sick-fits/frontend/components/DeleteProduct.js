import { useMutation } from "@apollo/client";
import gql from "graphql-tag";

const DELETE_PRODUCT_MUTATION = gql`
  mutation DELETE_PRODUCT_MUTATION($id: ID!) {
    deleteProduct(id: $id) {
      id
    }
  }
`;

function update(cache, payload) {
  cache.evict(cache.identify(payload.data.deleteProduct));
}

export function DeleteProduct({ id, children }) {
  const [deleteProduct, { loading }] = useMutation(DELETE_PRODUCT_MUTATION, {
    variables: { id },
    update,
  });
  function handleClick() {
    if (confirm("Are you sure you want to delete this item?")) {
      deleteProduct().catch((error) => alert(error.message));
    }
  }
  return (
    <button disabled={loading} type="button" onClick={handleClick}>
      {children}
    </button>
  );
}
