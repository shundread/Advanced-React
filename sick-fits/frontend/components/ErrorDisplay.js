export function ErrorDisplay({ error }) {
  return (
    <p>
      Error: <code>{error.message}</code>
    </p>
  );
}
