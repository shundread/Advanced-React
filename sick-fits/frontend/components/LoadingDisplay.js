export function LoadingDisplay() {
  return <p>Loading</p>;
}
