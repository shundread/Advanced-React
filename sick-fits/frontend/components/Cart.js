import { calculateCartTotalPrice } from "../lib/calculateCartTotalPrice";
import { formatMoney } from "../lib/formatMoney";
import { useUser } from "../lib/useUser";
import { CartItem } from "./CartItem";
import { CartStyles } from "./styles/CartStyles";
import { Supreme } from "./styles/Supreme";

export function Cart() {
  const user = useUser();
  if (!user) return null;
  return (
    <CartStyles open>
      <header>
        <Supreme>{user.name}</Supreme>
      </header>
      <ul>
        {user.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      <footer>
        <p>{formatMoney(calculateCartTotalPrice(user.cart))}</p>
      </footer>
    </CartStyles>
  );
}
