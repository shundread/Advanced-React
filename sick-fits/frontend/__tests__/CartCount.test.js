import { render, screen } from "@testing-library/react";
import wait from "waait";
import { CartCount } from "../components/CartCount";

describe("<CartCount />", () => {
  it("Renders", () => {
    const { container } = render(<CartCount count={10} />);
  });

  it("Matches the snapshot", () => {
    const { container } = render(<CartCount count={8} />);
    expect(container).toMatchSnapshot();
  });

  // There's an issue with this test where it depends on an animation concluding
  // so the tests will succeed. It's important to be careful here about usage of
  // expect(container.textContent).toBe instead of the more permissive test of
  // expect(container).toHaveTextContent since the latter will match in-transition
  // componets, while the former will complain about incomplete transitions
  it("Updates via props", async () => {
    const { container, rerender } = render(<CartCount count={11} />);
    expect(container.textContent).toBe("11"); // stricter test

    rerender(<CartCount count={12} />);

    // Wait for the transition to end
    await wait(400);
    expect(container.textContent).toBe("12"); // stricter test passes because of wait
  });
});
