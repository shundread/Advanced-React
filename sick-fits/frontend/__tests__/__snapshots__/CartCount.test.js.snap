// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CartCount /> Matches the snapshot 1`] = `
<div>
  <span
    class="CartCount__AnimationStyles-sc-xxvp4g-1 bjfODT"
  >
    <div>
      <div
        class="CartCount__Dot-sc-xxvp4g-0 dgpiUM count"
      >
        8
      </div>
    </div>
  </span>
</div>
`;
